### ISSUE TITLE
Plugin believes functions haven't been deployed, when in fact they have

### ISSUE THREAD

**Issue by mrcrilly** [matt-filion/serverless-external-s3-event#37]
**tl;dr**

I am unable to get this plugin to work as it cannot correctly identify that my functions have been deployed. The error from the plugin is as follows:

```
It looks like the function has not yet been deployed. You must use 'sls deploy' before doing 'sls s3deploy.
```

(there's an apostrophe missing from the end of that error message, by the way.)

The functions have indeed been deployed and exist, and work fine. The S3 buckets exist also, hence the need to use this plugin. After doing some debugging, I noticed the following issue at line `160` of `index.js` when it compares the deployed name versus the ARN name:

```
Serverless: ukrgdatapipeline-dev-ukrgdatapipeline versus ukrgdatapipeline-dev-UkRgDataPipeline
Serverless: ukrgdatapipeline-dev-RgToDocker versus ukrgdatapipeline-dev-UkRgDataPipeline
```

I've provided details at the bottom about my changes to produce this output.

It seems to only work with comparing one of the functions to all of the rest of them. That is, `ukrgdatapipeline-dev-UkRgDataPipeline` is the deployed name and it's used to compare against the ARNs of two functions to ensure they're deployed:

- `ukrgdatapipeline-dev-ukrgdatapipeline`
- `ukrgdatapipeline-dev-RgToDocker`

Either that or it's simply short circuiting.

Anyway it's not comparing the two correctly, and I'm unsure why until I do some further work.

Thoughts? More information provided below.

Version information:
Node.js: 8.60
Serverless: 1.23.0
serverless-plugin-existing-s3: 2.0.3

Here is my yaml template:

```
service:
  name: ukrgdatapipeline

plugins:
  - serverless-python-requirements
  - serverless-external-s3-event

provider:
  name: aws
  region: ap-southeast-2
  runtime: python2.7
  iamRoleStatements:
    - Effect: "Allow"
      Action:
        - "s3:*"
      Resource:
        - "arn:aws:s3:::${self:custom.UkDataS3Bucket}-${opt:stage}"
    - Effect: "Allow"
      Action:
        - "kms:*"
      Resource:
        - ${env:AWS_KMS_KEY_ARN}

custom:
  bamboo:
    project_id: ${env:BAMBOO_DEPLOYMENT_PROJECT_ID}
    environment_id: ${env:BAMBOO_DEPLOYMENT_ENVIRONMENT_ID}
    release_id: ${env:BAMBOO_DEPLOYMENT_RELEASE_ID}
  pythonRequirements:
      dockerizePip: true  
  ignoreme: ignoreme
functions:
  ukrgdatapipeline:
    name: <ignore me>
    runtime: <ignore me>
    memorysize: 512
    timeout: 240
    package:
      artifact: <ignore me>
    handler: <ignore me>
    description: Ignore Me
    environment:
      ETC: ETC
    events:
      - existingS3:
          bucket: ${self:custom.UkDataS3Bucket}-${opt:stage}
          events:
            - s3:ObjectCreated:*
          rules:
            - prefix: RouteingGuide/
            - suffix: .ZIP
  
  RgToDocker:
    name: ${self:service}-${opt:stage}-RgToDocker
    description: <ignore me>
    handler: <ignore me>
    runtime: python2.7
    environment:
      ETC: ETC
    events:
      - existingS3:
          bucket: ${self:custom.ProcessedUkDataS3Bucket}-${opt:stage}
          event: s3:ObjectCreated:*
          rules:
            - prefix: RouteingGuideData/
```

SLS_DEBUG message:

```
Serverless: Invoke aws:info
Service Information
service: ukrgdatapipeline
stage: dev
region: ap-southeast-2
stack: ukrgdatapipeline-dev
api keys:
  None
endpoints:
  None
functions:
  ukrgdatapipeline: ukrgdatapipeline-dev-ukrgdatapipeline
  RgToDocker: ukrgdatapipeline-dev-RgToDocker
Serverless: ukrgdatapipeline-dev-ukrgdatapipeline versus ukrgdatapipeline-dev-UkRgDataPipeline
Serverless: ukrgdatapipeline-dev-RgToDocker versus ukrgdatapipeline-dev-UkRgDataPipeline

  Error --------------------------------------------------

  It looks like the function has not yet been deployed. You must use 'sls deploy' before doing 'sls s3deploy.

     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.

  Stack Trace --------------------------------------------

Error: It looks like the function has not yet been deployed. You must use 'sls deploy' before doing 'sls s3deploy.
    at S3Deploy.getLambdaFunctionConfigurationFromDeployedStack (/Users/crilly/git/uk-rg-data-pipeline/deployment/serverless/node_modules/serverless-external-s3-event/index.js:166:13)
    at bucket.NotificationConfiguration.LambdaFunctionConfigurations.forEach (/Users/crilly/git/uk-rg-data-pipeline/deployment/serverless/node_modules/serverless-external-s3-event/index.js:232:36)
    at Array.forEach (<anonymous>)
    at bucketNotifications.forEach (/Users/crilly/git/uk-rg-data-pipeline/deployment/serverless/node_modules/serverless-external-s3-event/index.js:231:71)
    at Array.forEach (<anonymous>)
    at info.getStackInfo.then (/Users/crilly/git/uk-rg-data-pipeline/deployment/serverless/node_modules/serverless-external-s3-event/index.js:228:27)
From previous event:
    at S3Deploy.afterS3DeployFunctions (/Users/crilly/git/uk-rg-data-pipeline/deployment/serverless/node_modules/serverless-external-s3-event/index.js:222:32)
    at BbPromise.reduce (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:360:55)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:360:22)
    at PluginManager.run (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:391:17)
    at variables.populateService.then (/usr/local/lib/node_modules/serverless/lib/Serverless.js:99:33)
    at runCallback (timers.js:781:20)
    at tryOnImmediate (timers.js:743:5)
    at processImmediate [as _immediateCallback] (timers.js:714:5)
From previous event:
    at Serverless.run (/usr/local/lib/node_modules/serverless/lib/Serverless.js:86:74)
    at serverless.init.then (/usr/local/lib/node_modules/serverless/bin/serverless:39:50)
    at <anonymous>

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless

  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           8.6.0
     Serverless Version:     1.23.0

...

Serverless: Invoke s3deploy
Serverless: ukrgdatapipeline-dev-ukrgdatapipeline versus ukrgdatapipeline-dev-UkRgDataPipeline
Serverless: ukrgdatapipeline-dev-RgToDocker versus ukrgdatapipeline-dev-UkRgDataPipeline

  Error --------------------------------------------------

  It looks like the function has not yet been deployed. You must use 'sls deploy' before doing 'sls s3deploy.

     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.

  Stack Trace --------------------------------------------

Error: It looks like the function has not yet been deployed. You must use 'sls deploy' before doing 'sls s3deploy.
    at S3Deploy.getLambdaFunctionConfigurationFromDeployedStack (/Users/crilly/git/uk-rg-data-pipeline/deployment/serverless/node_modules/serverless-external-s3-event/index.js:166:13)
    at bucket.NotificationConfiguration.LambdaFunctionConfigurations.forEach (/Users/crilly/git/uk-rg-data-pipeline/deployment/serverless/node_modules/serverless-external-s3-event/index.js:232:36)
    at Array.forEach (<anonymous>)
    at bucketNotifications.forEach (/Users/crilly/git/uk-rg-data-pipeline/deployment/serverless/node_modules/serverless-external-s3-event/index.js:231:71)
    at Array.forEach (<anonymous>)
    at info.getStackInfo.then (/Users/crilly/git/uk-rg-data-pipeline/deployment/serverless/node_modules/serverless-external-s3-event/index.js:228:27)
From previous event:
    at S3Deploy.afterS3DeployFunctions (/Users/crilly/git/uk-rg-data-pipeline/deployment/serverless/node_modules/serverless-external-s3-event/index.js:222:32)
    at BbPromise.reduce (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:360:55)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:360:22)
    at PluginManager.run (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:391:17)
    at variables.populateService.then (/usr/local/lib/node_modules/serverless/lib/Serverless.js:99:33)
    at runCallback (timers.js:781:20)
    at tryOnImmediate (timers.js:743:5)
    at processImmediate [as _immediateCallback] (timers.js:714:5)
From previous event:
    at Serverless.run (/usr/local/lib/node_modules/serverless/lib/Serverless.js:86:74)
    at serverless.init.then (/usr/local/lib/node_modules/serverless/bin/serverless:39:50)
    at <anonymous>

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless

  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           8.6.0
     Serverless Version:     1.23.0
```

I added the following lines to `index.js` (line: `160`) to get some additional information about what's going on here:

**Original**:

```
let deployed = results.functions.find((fn) => fn.deployedName === cfg.LambdaFunctionArn);
```

**Edited**:

```
let deployed = results.functions.find((fn) => {
    this.serverless.cli.log(fn.deployedName + ' versus ' + cfg.LambdaFunctionArn);
    fn.deployedName === cfg.LambdaFunctionArn;
});
```

Which produces the following output, which can be seen above mixed into the console text:

```
Serverless: ukrgdatapipeline-dev-ukrgdatapipeline versus ukrgdatapipeline-dev-UkRgDataPipeline
Serverless: ukrgdatapipeline-dev-RgToDocker versus ukrgdatapipeline-dev-UkRgDataPipeline
```

**Comment by mrcrilly** [comment]
It seems to me as though the plugin can't handle a use case where multiple functions are being used and triggered on existing S3 buckets. In my case, each function need to trigger based on a separate, existing S3 bucket.

For the record, I've tried using `.toLowerCase()` to make the cases match, but the problem persists because it attempts to match the second function against the ARN of the first.

I'm still working on this for now.

**Comment by mrcrilly** [comment]
I was wrong about the plugin not being able to handle multiple functions. The problem seems to be here:

```
let deployed = results.functions.find((fn) => {
        this.serverless.cli.log(fn.deployedName + ' versus ' + cfg.LambdaFunctionArn);
        fn.deployedName === cfg.LambdaFunctionArn;
    });
```

(The `cli.log()` call was added by me to debug the situation.)

This call to `.find()` on the array isn't short circuiting when it finds a match:

```
Serverless: ukrgdatapipeline-dev-ukrgdatapipeline versus ukrgdatapipeline-dev-ukrgdatapipeline
Serverless: ukrgdatapipeline-dev-rgtodocker versus ukrgdatapipeline-dev-ukrgdatapipeline
```

It could be that there are difference in those string I can't see, such as invisible characters, but I doubt it, or something else is going on here.

**Comment by mrcrilly** [comment]
Aaand bingo. I've found the issue. Essentially it boils down to this: the name of the function as a key in the `functions` dictionary can't be different in any way to the name of the function via the `name` attribute. Example:

```
functions:
  testing01:
    name: Testing01
  
  testing02:
    name: testing02
```

Our testing01 function would cause the plugin to not see it has been deployed.

I'm not sure if this is a bug at this stage, or an error on my part. Should we be able to name the functions differently to their respective parent keys, or am I asking to eat my cake too? :-)

**Comment by k4200** [comment]
> Should we be able to name the functions differently to their respective parent keys, or am I asking to eat my cake too?

Since this plugin depends on AwsInfo plugin, and it returns only `name` and `deployedName`, I don't think we can use different function names from the keys of the functions dictionary.
https://github.com/serverless/serverless/blob/master/lib/plugins/aws/info/getStackInfo.js#L43-L45

**Comment by ocoulson** [comment]
@matt-filion I was having the same issue referenced above and the fix by @k4200 works perfectly. I've had to manually make the changes and commit the node-modules in my project because although I see the fix is merged in, I don't think it has been released, so the version I have from npm doesn't contain it. Maybe I'm looking in the wrong place, I saw from another issue that there are now 2 versions of the plugin. Would you mind making a release containing the fix above? Thanks a lot

**Comment by akaNightmare** [comment]
Still able to see the issue.
```
Serverless: Load command run
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command create
Serverless: Load command install
Serverless: Load command package
Serverless: Load command deploy
Serverless: Load command deploy:function
Serverless: Load command deploy:list
Serverless: Load command deploy:list:functions
Serverless: Load command invoke
Serverless: Load command invoke:local
Serverless: Load command info
Serverless: Load command logs
Serverless: Load command login
Serverless: Load command logout
Serverless: Load command metrics
Serverless: Load command print
Serverless: Load command remove
Serverless: Load command rollback
Serverless: Load command rollback:function
Serverless: Load command slstats
Serverless: Load command plugin
Serverless: Load command plugin
Serverless: Load command plugin:install
Serverless: Load command plugin
Serverless: Load command plugin:uninstall
Serverless: Load command plugin
Serverless: Load command plugin:list
Serverless: Load command plugin
Serverless: Load command plugin:search
Serverless: Load command emit
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command rollback
Serverless: Load command rollback:function
Serverless: Load command webpack
Serverless: Load command s3deploy
Serverless: Load command s3remove
Serverless: Invoke deploy
Serverless: Invoke package
Serverless: Invoke aws:common:validate
Serverless: Invoke aws:common:cleanupTempDir
Serverless: Invoke webpack:validate
Serverless: Invoke webpack:compile
Serverless: Bundling with Webpack...
Time: 2741ms
                            Asset     Size  Chunks             Chunk Names
            lambda/xml-handler.js   181 kB       0  [emitted]  lambda/xml-handler
lambda/schedule-alerts-handler.js  76.2 kB       1  [emitted]  lambda/schedule-alerts-handler
       lambda/checksum-handler.js  30.4 kB       2  [emitted]  lambda/checksum-handler
 lambda/checksum-daily-handler.js    30 kB       3  [emitted]  lambda/checksum-daily-handler
         lambda/alerts-handler.js  19.4 kB       4  [emitted]  lambda/alerts-handler
         lambda/stores-handler.js  19.6 kB       5  [emitted]  lambda/stores-handler
   [0] external "babel-runtime/regenerator" 42 bytes {0} {1} {2} {3} {4} {5} [not cacheable]
   [3] external "babel-runtime/core-js/promise" 42 bytes {0} {1} {2} {3} {4} {5} [not cacheable]
   [6] external "lodash.foreach" 42 bytes {0} {1} {2} {3} {4} {5} [not cacheable]
   [8] external "aws-sdk" 42 bytes {0} {1} {2} {3} {5} [not cacheable]
  [14] external "babel-runtime/core-js/object/assign" 42 bytes {0} {1} {4} [not cacheable]
  [39] ./lambda/xml-handler.js 3.84 kB {0} [built]
  [56] ./lambda/checksum-handler.js 1.88 kB {2} [built]
  [57] ./lambda/alerts-handler.js 11.3 kB {4} [built]
  [58] external "apn" 42 bytes {4} [not cacheable]
  [59] external "node-gcm" 42 bytes {4} [not cacheable]
  [60] ./services/device-token-cleaner.js 1.62 kB {4} [built]
  [61] ./lambda/schedule-alerts-handler.js 4.63 kB {1} [built]
  [66] ./lambda/stores-handler.js 1.86 kB {5} [built]
  [67] ./services/stores.js 10.2 kB {5} [built]
  [69] ./lambda/checksum-daily-handler.js 1.85 kB {3} [built]
    + 55 hidden modules
Serverless: Invoke webpack:package
Serverless: Package lock found - Using locked versions
Serverless: Packing external modules: babel-runtime@^6.26.0, lodash.foreach@^4.5.0, mongodb@3.0.2, lodash.uniq@^4.5.0, lodash.find@^4.6.0, xml2js@^0.4.19, moment@2.20.1, lodash.flatten@^4.4.0, lodash.omit@^4.5.0, lodash.filter@^4.6.0, lodash.union@^4.6.0, lodash.uniqwith@^4.5.0, lodash.isequal@^4.5.0, lodash.trim@^4.5.1, lodash.get@^4.4.2, lodash.mapvalues@^4.6.0, lodash.groupby@^4.6.0, lodash.clone@^4.5.0, redis@^2.8.0, lodash.isplainobject@^4.0.6, lodash.merge@^4.6.0, lodash.has@^4.5.2, lodash.reduce@^4.6.0, moment-timezone@^0.5.14, lodash.clonedeep@^4.5.0, lodash.compact@^3.0.1, lodash.keyby@^4.6.0, apn@^2.2.0, node-gcm@^0.14.9, lodash.pick@^4.4.0
Serverless: Packaging service...
Serverless: Invoke aws:package:finalize
Serverless: Invoke aws:common:moveArtifactsToPackage
Serverless: Invoke aws:common:validate
Serverless: Checking existing buckets actually exist
Serverless: WARNING: Inappropriate call of provider.request()
Serverless: All existing buckets actually exist
Serverless: Invoke aws:deploy:deploy
Serverless: Service files not changed. Skipping deployment...
Serverless: Invoke aws:info
Service Information
service: hbt-sls-xml-parser
stage: stage
region: us-east-1
stack: hbt-sls-xml-parser-stage
api keys:
  None
endpoints:
  None
functions:
  processXML: hbt-sls-xml-parser-stage-processXML
  checksum: hbt-sls-xml-parser-stage-checksum
  sendPushNotification: hbt-sls-xml-parser-stage-sendPushNotification
  dayPartsAlerts: hbt-sls-xml-parser-stage-dayPartsAlerts
  removeExpiredStores: hbt-sls-xml-parser-stage-removeExpiredStores
  checksumDaily: hbt-sls-xml-parser-stage-checksumDaily
 
  Error --------------------------------------------------
 
  It looks like the function has not yet been deployed. You must use 'sls deploy' before doing 'sls s3deploy.
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Stack Trace --------------------------------------------
 
Error: It looks like the function has not yet been deployed. You must use 'sls deploy' before doing 'sls s3deploy.
    at S3Deploy.getLambdaFunctionConfigurationFromDeployedStack (/Users/ivan.zubok/PhpstormProjects/heartbeat/hbt-sls-xmlparser/node_modules/serverless-external-s3-event/index.js:163:13)
    at bucket.NotificationConfiguration.LambdaFunctionConfigurations.forEach (/Users/ivan.zubok/PhpstormProjects/heartbeat/hbt-sls-xmlparser/node_modules/serverless-external-s3-event/index.js:229:36)
    at Array.forEach (<anonymous>)
    at bucketNotifications.forEach (/Users/ivan.zubok/PhpstormProjects/heartbeat/hbt-sls-xmlparser/node_modules/serverless-external-s3-event/index.js:228:71)
    at Array.forEach (<anonymous>)
    at info.getStackInfo.then (/Users/ivan.zubok/PhpstormProjects/heartbeat/hbt-sls-xmlparser/node_modules/serverless-external-s3-event/index.js:225:27)
From previous event:
    at S3Deploy.afterS3DeployFunctions (/Users/ivan.zubok/PhpstormProjects/heartbeat/hbt-sls-xmlparser/node_modules/serverless-external-s3-event/index.js:219:32)
    at BbPromise.reduce (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:368:55)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:368:22)
    at PluginManager.run (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:399:17)
    at variables.populateService.then (/usr/local/lib/node_modules/serverless/lib/Serverless.js:102:33)
    at runCallback (timers.js:756:18)
    at tryOnImmediate (timers.js:717:5)
    at processImmediate [as _immediateCallback] (timers.js:697:5)
From previous event:
    at Serverless.run (/usr/local/lib/node_modules/serverless/lib/Serverless.js:89:74)
    at serverless.init.then (/usr/local/lib/node_modules/serverless/bin/serverless:42:50)
    at <anonymous>
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless
 
  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           9.4.0
     Serverless Version:     1.25.0
```

**Comment by jieren172** [comment]
In my tests, this issue does not appear in version 1.0.9 but appear in 1.0.10

**Comment by docinfosci** [comment]
I just realized that an alternate version of this plugin has been released: https://www.npmjs.com/package/serverless-plugin-existing-s3

**Comment by gitgoose** [comment]
@docinfosci @jieren172 @matt-filion 
Well i am using latest and still have the issue
Using python(3.6) based serverless app with sls version 1.26.1
serverless-plugin-existing-s3 (v 2.1.0)
Help appreciated!!


$ sls s3deploy
Serverless: beforeFunctions --> building ... 
Serverless: beforeFunctions <-- Complete, built 1 events.
Serverless: functions --> prepare to be executed by s3 buckets ... 
policyId exS3-v2-<lambda-name>-<bucket-name>
Serverless: functions <-- **built 0 events across 1 buckets.** 
Serverless: beforeS3 --> 
Serverless: beforeS3 <-- 
Serverless: s3 --> initiate requests ...
 
  Type Error ---------------------------------------------
 
  eventTypes.join is not a function

I do this step after
sls deploy 
with the following in my resources section[Not sure if this is even needed]
resources:
  Resources:
    LambdaInvokePermission:
      Type: AWS::Lambda::Permission
      DependsOn:
          - InfraLambdaFunction
      Properties:
        Action: lambda:InvokeFunction
        FunctionName: <name>-${self:custom.currentStage}
        Principal: s3.amazonaws.com
        SourceArn: arn:aws:s3:::${file(configs/config.${self:custom.currentStage}.json):s3Bucket}

**Comment by matt-filion** [comment]
What npm and node version? .join not being a function leads me to believe
your node version is out of date.

**Comment by matt-filion** [comment]
stale issue.
